server/src/actors/events.rs

```diff
@@ -125,6 +132,172 @@ impl EventProcessor {
+    fn event_from_minidump_data(
+        form_data: Item,
+        event_data_mps: Option<Item>,
+        breadcrumbs1: Option<Item>,
+        breadcrumbs2: Option<Item>,
+    ) -> Annotated<Event> {
+        // if we have a "sentry" entry it should be a straight json.. merge this into
+        if form_data.ty() != ItemType::FormData
+            || form_data.content_type() != Some(&ContentType::Text)
+        {
+            //this should never happen caller called it with
+            log::error!("merge_form_data called without a proper FormData item");
+            return Annotated::new(Event::default());
+        }
+
+        #[derive(serde::Deserialize)]
+        struct StringPair<'a>(pub &'a str, pub &'a str);
+
+        let payload = form_data.payload();
+        let mut consolidated_data_stream =
+            serde_json::Deserializer::from_slice(payload.deref()).into_iter();
+
+        let mut event_data = SerdeValue::Object(SerdeMap::new());
+        loop {
+            match consolidated_data_stream.next() {
+                None => break,
+                Some(Err(e)) => log::error!("Form data deserialization failed {:?}", e),
+                Some(Ok(val)) => {
+                    let val: StringPair = val;
+                    let (key, val) = (val.0, val.1);
+                    if let Some(keys) = get_sentry_entry_indexes(key) {
+                        //iterate through all 'sentry[...' params
+                        update_json_object(&mut event_data, &keys, val);
+                    } else if key == "sentry" {
+                        //this should be a json string representing the event, merge it into what
+                        //we have already created
+                        let evt_from_json = serde_json::from_str(val);
+                        if let Ok(evt_from_json) = evt_from_json {
+                            merge_vals(&mut event_data, evt_from_json)
+                        }
+                    } else {
+                        //an unknown entry, just add it to extra
+                        update_json_object(&mut event_data, &["extra", key], val);
+                    }
+                }
+            }
+        }
+
+        if let Some(event_data_mps) = event_data_mps {
+            if event_data_mps.len() < MAX_MSGPACK_EVENT_SIZE_BYTES {
+                if let Ok(evt) = mps::from_slice(event_data_mps.payload().as_ref()) {
+                    merge_vals(&mut event_data, evt);
+                } else {
+                    log::debug!("Invalid message pack event data in form data");
+                }
+            }
+        }
+
+        let mut annotated_event = Annotated::deserialize_with_meta(event_data).unwrap_or_default();
+
+        Self::add_message_pack_breadcrumbs(&mut annotated_event, breadcrumbs1, breadcrumbs2);
+
+        Self::write_minidump_placeholder(&mut annotated_event);
+
+        annotated_event
+    }
+
+    fn get_item_by_name(envelope: &mut Envelope, name: &str) -> Option<Item> {
```

Comment:
Should be a method on `Envelope` or a global func. Also it's not obvious from the func name that this method removes items.

Reply:
ok, renamed to take_item_by_name and moved it outside Envelope